extensions/src/main/java/org/pentaho/platform/plugin/services/email/EmailService.java

```diff
@@ -186,18 +237,37 @@ protected PasswordAuthentication getPasswordAuthentication() {
       msg.setText( messages.getString( "EmailService.MESSAGE" ) );
       msg.setHeader( "X-Mailer", "smtpsend" );
       msg.setSentDate( new Date() );
-      Transport.send( msg );
-      sendEmailMessage = "EmailTester.SUCESS";
+      sendEmail( emailConfig, token, session, msg );
+      sendEmailMessage = TEST_EMAIL_SUCCESS;
     } catch ( Exception e ) {
       logger.error( messages.getString( "EmailService.NOT_CONFIGURED" ), e );
-      sendEmailMessage = "EmailTester.FAIL";
+      sendEmailMessage = TEST_EMAIL_FAIL;
     }
+    setEmailConfig( emailConfig );
     return sendEmailMessage;
   }
 
+  public void sendEmail( final IEmailConfiguration emailConfig, IEmailAuthenticationResponse token, Session session, MimeMessage msg ) throws Exception {
+    if ( emailConfig.getAuthMechanism().equals( XOAUTH ) ) {
+      if ( emailConfig.getSmtpProtocol().equalsIgnoreCase( "graph_api" ) ) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        msg.writeTo( os );
+        String s = Base64.getEncoder().encodeToString( os.toByteArray() );
+        sendMailGraphApi( emailConfig, token.getAccessToken(), s );
+      } else {
+        SMTPTransport transport = new SMTPTransport( session, null );
+        transport.connect( emailConfig.getSmtpHost(), emailConfig.getUserId(), null );
+        transport.issueCommand( OAUTH_COMMAND + Base64Utils.toBase64( String.format( USER_TOKEN_STRING, emailConfig.getUserId(), token.getAccessToken() ).getBytes() ), 235 );
+        transport.sendMessage( msg, msg.getAllRecipients() );
```

Comment:
This is still a bit hard to read.

About the magic 235 number... According to https://en.wikipedia.org/wiki/List_of_SMTP_server_return_codes, that's the HTTP/SMTP status code for "Authentication Succeeded". So, could you create a constant for it, e.g.:
```java
private static final int SMTP_SC_AUTH_SUCCESS = 225;
```

Also, could you please create a local variable holding the command string?
For example:
```java
String userTokenString = String.format( USER_TOKEN_STRING, emailConfig.getUserId(), token.getAccessToken() );
String smtpCommand = OAUTH_COMMAND + Base64Utils.toBase64( userTokenString.getBytes() );

// ...

transport.issueCommand( smtpCommand, SMTP_SC_AUTH_SUCCESS );
```

Finally, do the two above Base 64 encoding methods really work differently?
`Base64.getEncoder().encodeToString(.)` vs `Base64Utils.toBase64( . )`.